src/main/java/org/seqdoop/hadoop_bam/BAMRecordReader.java

```diff
@@ -224,25 +232,23 @@ public static long getKey0(int refIdx, int alignmentStart0) {
 		return false;
 	}
 
-	private static boolean overlaps(SAMRecord r, List<Locatable> intervals) {
+	private boolean overlaps(SAMRecord r) {
 		if (intervals == null ||
 				(r.getReadUnmappedFlag() && r.getAlignmentStart() == SAMRecord.NO_ALIGNMENT_START)) {
 			return true;
 		}
-		for (Locatable interval : intervals) {
-			if (r.getReadUnmappedFlag()) {
+		if (r.getReadUnmappedFlag()) { // special case for unmapped reads with coordinate set
+			for (Locatable interval : intervals) {
 				if (interval.getStart() <= r.getStart() && interval.getEnd() >= r.getStart()) {
```

Comment:
One meta comment - I think the intervals we're getting from the user are intended to be [query intervals](https://github.com/samtools/htsjdk/blob/c847e260a351dd4ded3caf3c44c6ec7d0c037675/src/java/htsjdk/samtools/QueryInterval.java#L12), which means the interval end value can use 0 or -1 as a sentinel to mean the end of the reference sequence.  So this code path, for example, should return true for an unmapped, placed read located at [20, x] against the query interval [10, -1]. It doesn't look to me like the overlap detector is intended to handle these either. We should probably add tests for these cases.

The code in htsjdk for handling these is a bit distributed, but there is code in the QueryInterval class as well as [here](https://github.com/samtools/htsjdk/blob/dfe2b7873beec5ce1bc9ebb410dbc6f05f6bbcda/src/java/htsjdk/samtools/BAMFileReader.java#L944) that illustrates how BAM index queries handle this.